README.md

```diff
@@ -60,194 +60,86 @@ addTen(2); // 12
-        for (let i = 100; i > 0; i -= 10) {
-            fadeOut(i);
-            await delay(10);
-        }
-      }
-  }
-  ```
-* Given `g = f(?)`, the `this` receiver of the function `f` is fixed as `undefined` in the partially
-  applied function result `g`:
-  ```js
-  function f(x) { return `this: ${this}, x: ${x}`; }
-  const o = { g: f(?) };
-  o.g(2); // 'this: undefined, x: 2'
-  ```
-  However, you may uncurry `this` using `f.call`:
-  ```js
-  function f(x) { return `this: ${this}, x: ${x}.`; }
-  const g = f.call(?, 2);
-  g(1, 2); // `this: 1, x: 2`
-  ```
-* Given `g = o.f(?)`, the `this` receiver of the function `o.f` is fixed as `o` in the partially 
-  applied function result `g`:
-  ```js
-  const o = { f(x) { return `this.y: ${this.y}, x: ${x}`; }, y: 1 };
-  const g = o.f(?);
-  g(2); // 'this.y: 1, x: 2'
-  ```
-  However, you may uncurry `this` using `o.f.call`:
-  ```js
-  const o = { f(x) { return `this.y: ${this.y}, x: ${x}`; }, y: 1 };
-  const g = o.f.call(?, 3);
-  g({ y: 4 }); // 'this.y: 4, x: 3'
-  ```
-* Given `g = new f(?)`, the partially applied function result `g` is a function that when called will 
-  construct a new instance of `f`:
-  ```js
-  function f(x, y) { this.z = `${x}, ${y}` }
-  const g = new f("a", ?);
-  const obj = g(1); // creates an f instance
-  obj.z; // 'a, 1'
-  ```
-* Given `g = f(?)`, the `length` of the partially applied function result `g` is equal to the number of `?` placeholder tokens in the
-  argument list:
-  ```js
-  const f = (x, y) => x + y;
-  const g = f(?, 2);
-  f.length; // 2
-  g.length; // 1
-  ```
+In both of the above examples, excess supplied arguments to `g` are ignored and **not** passed on 
+to the partially applied function. The ability to pass on additional arguments is not part of this
+proposal but may be considered in a future proposal.
+
+The following is a list of additional semantic rules:
+
+* Given `f(?)`, the expression `f` is evaluated immediately.
+* Given `f(?, x)`, the non-placeholder argument `x` is evaluated immediately and fixed in its position.
+* Given `f(?)`, excess arguments supplied to the partially applied function result are ignored.
+* Given `f(?, ?)` the partially applied function result will have a parameter for each placeholder 
+  token that is supplied in that token's position in the argument list.
+* Given `f(this, ?)`, the `this` in the argument list is the lexical `this`.
+* Given `f(?)`, the `this` receiver of the function `f` is fixed as `undefined` in the partially
+  applied function result.
+* Given `f(?)`, the `length` of the partially applied function result is equal to the number of `?` placeholder tokens in the
+  argument list.
+* Given `f(?)`, the `name` of the partially applied function result is `f.name`.
+* Given `o.f(?)`, the references to `o` and `o.f` are evaluated immediately.
+* Given `o.f(?)`, the `this` receiver of the function `o.f` is fixed as `o` in the partially 
```

Comment:
Now that we're getting away from the lazy eval semantics, I want to argue against this point specifically. Allowing the call to define the `this` context allows a bit more generic use. Taking form the underscore codebase:

```js
// https://github.com/jashkenas/underscore/blob/5c237a7c/underscore.js#L948-L962
// setup...
_.before = function(times, func) {
  var memo;
  return function() {
    if (--times > 0) {
      memo = func.apply(this, arguments);
    }
    if (times <= 1) func = null;
   return memo;
  };
};
_.once = _.partial(_.before, 2);

// With partial application
// This also needs to ferry this context, and all additional args
_.once = _.before(2, ?...);

// use
const obj = {
  method: _.once(function(...args) {
    // this === obj
    // ...
  });
};
```

For a general purpose helper like `once`, we can't know the `this` context beforehand. I certainly wouldn't expect it to be `_` (or `undefined`, if we were to extract it outside the `_` global object).

Instead, we could rely on the bind syntax operator `::` (is that ever going to come back?), or explicitly call `Function.p.bind` after the partial application:

```js
const f = obj.foo(?).bind(obj);

f(1);
```

<details>

<summary>Weaker use case</summary>

```js
// https://github.com/jashkenas/underscore/blob/5c237a7c/underscore.js#L912-L917
// setup...
 _.wrap = function(func, wrapper) {
    return _.partial(wrapper, func);
  };

// With partial application
_.wrap = function(func, wrapper) {
  // We need to ferry both `this` context, and all additional args
  return wrapper(func, ?...);
};

// use
const obj = {
  method() {
    // ...
  }
}
obj.method = _.wrap(obj.method, function(orig, ...args) {
  // pre
  // this === obj
  const ret = orig.apply(this, args);
  // post
  return ret;
})
```

</details>

Reply:
> ```js
> // This also needs to ferry this context, and all additional args
> _.once = _.before(2, ?...);
> 
> // use
> const obj = {
>   method: _.once(function(...args) {
>     // this === obj
>     // ...
>   });
> };
> ```

This example actually works as you expect it to, since we bind the `this` of `once` to `_`, but both `once` *and* `before` return a new `function` when called that does not have a bound `this`.

One of the goals of this proposal is that these two statements are functionally equivalent:

```js
// partial application
obj.method(?, 1)("a")

// normal execution
obj.method("a", 1);
```

This is because we evaluate everything ahead of time *except* for the missing arguments, including determing the receiver up front.

Fixing the receiver hasn't changed between the original proposal and this PR. That said, I'm open to discussing possible syntax additions that could relax this behavior in the future.

Reply:
> both once and before return a new function when called that does not have a bound this.

Lol, don't know how I missed this. With that clarified, I have no objection.